File: ClaudeRelayAndroid/core-session/src/main/kotlin/relay/session/RecoveryController.kt
package relay.session

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlin.coroutines.cancellation.CancellationException

/**
 * Owns the recovery state machine ported from
 * Sources/ClaudeRelayClient/ViewModels/RecoveryController.swift: the
 * auto-suspend circuit breaker, generation tokens, cooldown tracking, and the
 * `handleForegroundTransition` / `restoreSession` flow.
 *
 * This is a faithful behavioral rewrite of the Swift source, not the plan
 * skeleton. Everything the Swift original reaches through its coordinator /
 * connection is injected as a lambda so the controller is testable without any
 * Android dependency. The nine load-bearing behaviors verified against the
 * source:
 *
 *  1. Backoff `[0,1,2,4,8,15]` seconds (RecoveryController.swift:176).
 *  2. The [isAlive] probe wraps a 5 s pong wait (RelayConnection.swift:313);
 *     this controller just calls [isAlive] and trusts that timeout.
 *  3. isAlive short-circuit: alive -> fetch + return WITHOUT entering the
 *     backoff loop; `isRecovering` is set true only AFTER the alive==false
 *     branch (RecoveryController.swift:151-164), so the alive path never
 *     flashes recovery UI.
 *  4. Auth/resume failure is TERMINAL: only RECONNECT failures retry the
 *     backoff array. A successful reconnect whose reauth/resume fails accounts
 *     the failure ONCE and returns — it does NOT re-run the loop
 *     (RecoveryController.swift:202-206, 272).
 *  5. App-level vs transport-level split: an app-level restore error sets
 *     [sessionAttachFailed], does NOT set [connectionTimedOut], but still
 *     counts toward the breaker (RecoveryController.swift:258-272).
 *  6. Circuit breaker: 3 consecutive AUTO failures -> [autoRecoverySuspended];
 *     only AUTO (not user) failures count; reset by user signals and by
 *     [resetAutoRecoveryBreaker] (RecoveryController.swift:47, 69-74,
 *     128-129, 224-233).
 *  7. 3 s auto-recovery cooldown: [scheduleAutoRecovery] drops auto-recoveries
 *     within 3 s of the last one ENDING (RecoveryController.swift:43, 94-98),
 *     stamping `lastRecoveryEndedAt` from the injected [nowMs] clock in the
 *     `finally` (RecoveryController.swift:169).
 *  8. [resetAutoRecoveryBreaker] clears suspension + failures
 *     (RecoveryController.swift:69-74); wired to a healthy-ping signal.
 *  9. [cancel] suspends the breaker, sets a recovery-failed flag, stamps
 *     `lastCancelledAt`, and bumps the generation so in-flight recovery bails
 *     (RecoveryController.swift:286-299); [triggerUserRecovery] ignores calls
 *     within 1 s of a cancel (RecoveryController.swift:56-58, 124-127).
 *
 * Plus: `recoveryGeneration` captured at entry and rechecked at every await
 * checkpoint (bail if newer); the two distinct guards `isRecoveryDispatched`
 * (sync entry-lock) vs [isRecovering] (UI StateFlow); [suppressSends] toggled
 * around the recovery body; and defer-idempotency — a single try/finally clears
 * `isRecovering` + `suppressSends(false)` + `isRecoveryDispatched` + stamps
 * `lastRecoveryEndedAt` even on a mid-await cancel.
 *
 * Not thread-safe across dispatchers — the Swift original is `@MainActor`
 * isolated. Callers must confine entry points to a single (main) dispatcher;
 * the [isRecoveryDispatched] sync entry-lock relies on that confinement.
 *
 * @param scope The coroutine scope recovery passes launch on. It MUST use a
 *   serial / confined dispatcher (e.g. `Dispatchers.Main.immediate`) — the
 *   [isRecoveryDispatched] sync entry-lock and the single-flight generation
 *   guards assume entry points and the recovery body never run concurrently,
 *   faithful to the Swift `@MainActor` model. The Task-6 coordinator wiring is
 *   responsible for providing such a scope.
 * @param isAlive Probes the socket; wraps a 5 s pong wait at the call site.
 * @param reconnect Forces a transport reconnect; throws on failure (only this
 *   failure retries the backoff array).
 * @param reauth Resets + re-establishes auth on the current connection.
 * @param resumeActive Resumes the active session (no-op if there is none).
 * @param fetchSessions Refreshes the server session list.
 * @param suppressSends Toggled true around the recovery body, false on exit.
 * @param isApplicationLevelError Classifies a restore error: app-level (session
 *   gone) vs transport-level. App-level errors set [sessionAttachFailed]
 *   instead of [connectionTimedOut].
 * @param sleepMs Backoff sleep; injectable for virtual-time tests.
 * @param nowMs Clock for cooldown/debounce stamps; injectable so tests control
 *   time. The default `System.currentTimeMillis()` is WALL-CLOCK and
 *   non-monotonic — an NTP correction or manual clock change can move it
 *   backwards and corrupt the 3 s cooldown / 1 s cancel-debounce windows. The
 *   Task-6 coordinator wiring MUST inject the Android-correct monotonic source
 *   `android.os.SystemClock.elapsedRealtime()` instead. This module stays
 *   pure-JVM (no Android dependency), so the monotonic clock is injected from
 *   the Android layer rather than referenced here. Never stamped as 0L.
 */
class RecoveryController(
    private val scope: CoroutineScope,
    private val isAlive: suspend () -> Boolean,
    private val reconnect: suspend () -> Unit,
    private val reauth: suspend () -> Unit,
    private val resumeActive: suspend () -> Unit,
    private val fetchSessions: suspend () -> Unit,
    private val suppressSends: (Boolean) -> Unit = {},
    private val isApplicationLevelError: (Throwable) -> Boolean = { false },
    private val sleepMs: suspend (Long) -> Unit = { delay(it) },
    private val nowMs: () -> Long = { System.currentTimeMillis() },
) {
    private companion object {
        /** Backoff delays in seconds (RecoveryController.swift:176). */
        val BACKOFF_SECONDS = longArrayOf(0, 1, 2, 4, 8, 15)

        /** Auto-recovery cooldown, ms (RecoveryController.swift:43 — 3 s). */
        const val AUTO_RECOVERY_COOLDOWN_MS = 3_000L

        /** Max consecutive auto failures before suspend (RecoveryController.swift:47). */
        const val MAX_AUTO_RECOVERY_FAILURES = 3

        /** Cancel debounce for user recovery, ms (RecoveryController.swift:124 — 1 s). */
        const val CANCEL_DEBOUNCE_MS = 1_000L
    }

    // MARK: - UI state (StateFlows)

    private val _isRecovering = MutableStateFlow(false)
    /** True while the backoff/restore body is running (drives recovery UI). */
    val isRecovering: StateFlow<Boolean> = _isRecovering.asStateFlow()

    private val _phase = MutableStateFlow(RecoveryPhase.RECONNECTING)
    val phase: StateFlow<RecoveryPhase> = _phase.asStateFlow()

    private val _connectionTimedOut = MutableStateFlow(false)
    /** Set on a terminal transport-level failure; tears the workspace down. */
    val connectionTimedOut: StateFlow<Boolean> = _connectionTimedOut.asStateFlow()

    private val _sessionAttachFailed = MutableStateFlow(false)
    /** Set on an app-level restore failure (session gone); connection stays up. */
    val sessionAttachFailed: StateFlow<Boolean> = _sessionAttachFailed.asStateFlow()

    private val _autoRecoverySuspended = MutableStateFlow(false)
    /** True when auto-retry is circuit-broken; awaits a user/healthy-ping signal. */
    val autoRecoverySuspended: StateFlow<Boolean> = _autoRecoverySuspended.asStateFlow()

    // MARK: - Breaker + generation state

    /**
     * Monotonic token bumped at the start of every recovery pass and on
     * [cancel]. A pass only proceeds while the token it captured still matches —
     * prevents a stale pass from restoring after a newer one or a cancel
     * (RecoveryController.swift:35, 158-159).
     */
    private var recoveryGeneration: ULong = 0u

    /**
     * Last recovery completion stamp (success or failure); cooldown anchor.
     * Null means "no recovery has ended yet" (Swift `Date.distantPast`), so the
     * first auto-recovery is always allowed — avoids the `now - MIN_VALUE`
     * overflow a sentinel Long would cause.
     */
    private var lastRecoveryEndedAt: Long? = null

    /** Consecutive auto-triggered recovery failures (RecoveryController.swift:41). */
    private var consecutiveAutoRecoveryFailures = 0

    /** Recovery-failed flag (Swift `coordinator.recoveryFailed`). Internal — the
     * UI binds to the more specific [connectionTimedOut] / [sessionAttachFailed]. */
    private var recoveryFailed = false

    /**
     * Synchronous entry lock, distinct from [isRecovering] (which is set after
     * `await isAlive()`). Prevents concurrent dispatches from racing across the
     * suspension at the top of the flow (RecoveryController.swift:55, 144).
     */
    private var isRecoveryDispatched = false

    /**
     * Stamp set by [cancel]; gates [triggerUserRecovery]
     * (RecoveryController.swift:58). Null means "never cancelled" so user
     * recovery is not debounced until a cancel actually happens.
     */
    private var lastCancelledAt: Long? = null

    /** The in-flight recovery job, so [cancel] can cancel it. */
    private var recoveryJob: Job? = null

    // MARK: - Breaker

    /**
     * Clears the auto-recovery circuit breaker. No-op when already idle so
     * healthy steady-state traffic doesn't thrash (RecoveryController.swift:69-74).
     * Wired to a healthy-ping signal by the coordinator.
     */
    fun resetAutoRecoveryBreaker() {
        if (!_autoRecoverySuspended.value && consecutiveAutoRecoveryFailures == 0) return
        consecutiveAutoRecoveryFailures = 0
        _autoRecoverySuspended.value = false
    }

    /**
     * Clears the terminal recovery flags ([connectionTimedOut] /
     * [sessionAttachFailed]). These are set-only by the recovery flow; the
     * coordinator/UI resets them on a fresh foreground transition or when the
     * user dismisses the recovery error. Mirrors the Swift reset on WorkspaceView
     * / SharedSessionCoordinator.
     */
    fun clearTerminalFlags() {
        _connectionTimedOut.value = false
        _sessionAttachFailed.value = false
    }

    // MARK: - Entry points

    /**
     * Auto-triggered recovery (from a send/ping failure). Gated by:
     * already-dispatched, auto-suspend (breaker open), and a 3 s cooldown since
     * the last recovery ended (RecoveryController.swift:84-104).
     */
    fun scheduleAutoRecovery() {
        if (isRecoveryDispatched) return
        if (_autoRecoverySuspended.value) return
        val endedAt = lastRecoveryEndedAt
        if (endedAt != null && nowMs() - endedAt < AUTO_RECOVERY_COOLDOWN_MS) return
        isRecoveryDispatched = true
        recoveryJob = scope.launch { handleForegroundTransition(userInitiated = false) }
    }

    /**
     * Explicit user-initiated recovery (foreground, network restored, rescan).
     * Clears the breaker so auto-retry resumes after this attempt. Ignores calls
     * within 1 s of a [cancel] to avoid sheet-dismiss -> foreground -> re-trigger
     * (RecoveryController.swift:118-134).
     */
    fun triggerUserRecovery() {
        if (isRecoveryDispatched) return
        val cancelledAt = lastCancelledAt
        if (cancelledAt != null && nowMs() - cancelledAt < CANCEL_DEBOUNCE_MS) return
        _autoRecoverySuspended.value = false
        consecutiveAutoRecoveryFailures = 0
        isRecoveryDispatched = true
        recoveryJob = scope.launch { handleForegroundTransition(userInitiated = true) }
    }

    /**
     * Test seam / direct entry: launches a recovery pass on [scope] bypassing the
     * scheduling gates, matching the Swift `coordinator.recoveryTask = Task { ... }`
     * dispatch. Sets the dispatch lock so re-entry is blocked, just like the real
     * entry points.
     */
    internal fun launchRecovery(userInitiated: Boolean) {
        if (isRecoveryDispatched) return
        isRecoveryDispatched = true
        recoveryJob = scope.launch { handleForegroundTransition(userInitiated) }
    }

    // MARK: - Recovery flow

    /**
     * The recovery pass (Swift `handleForegroundTransition`). Outer `finally`
     * clears the sync dispatch lock for ALL paths (including the alive
     * short-circuit), mirroring the outermost Swift `defer` at :144. The inner
     * `finally` — entered only on the not-alive path — clears the recovery UI
     * state and stamps the cooldown, mirroring the inner Swift `defer` at :166.
     */
    private suspend fun handleForegroundTransition(userInitiated: Boolean) {
        try {
            if (_isRecovering.value) return // already recovering (RecoveryController.swift:146)

            // isAlive short-circuit (RecoveryController.swift:151-156). isRecovering
            // is NOT set yet, so this path never flashes recovery UI.
            if (isAlive()) {
                fetchSessions()
                return
            }

            // Not alive: now we commit to a recovery pass.
            recoveryGeneration += 1u
            val myGeneration = recoveryGeneration

            _phase.value = RecoveryPhase.RECONNECTING
            recoveryFailed = false
            _isRecovering.value = true
            suppressSends(true)
            try {
                runReconnectThenRestore(myGeneration, userInitiated)
            } finally {
                // Inner defer (RecoveryController.swift:166-170): idempotent even
                // on a mid-await cancel. Generation-guarded so a stale gen-N
                // coroutine's finally can never clobber a mid-flight gen-(N+1)'s
                // freshly-set state. Under the current serial/confined dispatcher
                // (faithful to Swift's @MainActor) this guard is a no-op — gen-N's
                // body fully unwinds before gen-(N+1) is dispatched — but it is
                // cheap insurance against a future re-entrant ordering change.
                // The OUTER `isRecoveryDispatched` clear stays ungated (matches
                // Swift, and gating it would risk stranding the dispatch lock).
                if (myGeneration == recoveryGeneration) {
                    _isRecovering.value = false
                    suppressSends(false)
                    lastRecoveryEndedAt = nowMs()
                }
            }
        } finally {
            // Outer defer (RecoveryController.swift:144).
            isRecoveryDispatched = false
        }
    }

    /**
     * The backoff reconnect loop + terminal restore. Only [reconnect] failures
     * retry the backoff array (RecoveryController.swift:178-210); once
     * reconnected we hand off to [restoreSession], whose failure is TERMINAL.
     */
    private suspend fun runReconnectThenRestore(myGeneration: ULong, userInitiated: Boolean) {
        var reconnected = false
        for (attempt in BACKOFF_SECONDS.indices) {
            val delaySeconds = BACKOFF_SECONDS[attempt]
            // Generation/cancellation checkpoint (RecoveryController.swift:179).
            if (myGeneration != recoveryGeneration || !currentlyActive()) return

            if (delaySeconds > 0) {
                try {
                    sleepMs(delaySeconds * 1_000L)
                } catch (_: CancellationException) {
                    return // cancelled during backoff (RecoveryController.swift:186-189)
                }
                if (myGeneration != recoveryGeneration) return
            }

            try {
                reconnect()
                reconnected = true
                break
            } catch (_: CancellationException) {
                return // cancelled during reconnect (RecoveryController.swift:197-199)
            } catch (_: Throwable) {
                if (attempt == BACKOFF_SECONDS.size - 1) {
                    // Last attempt failed -> terminal transport failure
                    // (RecoveryController.swift:202-207).
                    if (myGeneration != recoveryGeneration) return
                    recoveryFailed = true
                    _connectionTimedOut.value = true
                    recordAutoRecoveryOutcome(success = false, userInitiated = userInitiated)
                    return
                }
                // Otherwise fall through to the next backoff attempt.
            }
        }

        if (!reconnected || myGeneration != recoveryGeneration) return
        restoreSession(myGeneration, userInitiated)
    }

    /**
     * Records a recovery outcome and updates the circuit breaker
     * (RecoveryController.swift:217-234). Only AUTO (not user) failures count.
     */
    private fun recordAutoRecoveryOutcome(success: Boolean, userInitiated: Boolean) {
        if (success) {
            consecutiveAutoRecoveryFailures = 0
            _autoRecoverySuspended.value = false
            return
        }
        if (userInitiated) return
        consecutiveAutoRecoveryFailures += 1
        if (consecutiveAutoRecoveryFailures >= MAX_AUTO_RECOVERY_FAILURES) {
            _autoRecoverySuspended.value = true
        }
    }

    /**
     * Restores auth + the active session on the freshly-reconnected transport
     * (RecoveryController.swift:237-280). Failure here is TERMINAL — it accounts
     * the outcome ONCE and returns; it never re-enters the backoff loop.
     */
    private suspend fun restoreSession(generation: ULong, userInitiated: Boolean) {
        _phase.value = RecoveryPhase.AUTHENTICATING
        try {
            reauth()
            if (generation != recoveryGeneration) return
            _phase.value = RecoveryPhase.RESUMING
            resumeActive()
            if (generation != recoveryGeneration) return
        } catch (_: CancellationException) {
            return // restore cancelled (RecoveryController.swift:250-252)
        } catch (error: Throwable) {
            if (generation != recoveryGeneration) return
            recoveryFailed = true
            // App-level errors leave the connection intact; transport-level
            // errors time the connection out (RecoveryController.swift:257-271).
            if (isApplicationLevelError(error)) {
                _sessionAttachFailed.value = true
            } else {
                _connectionTimedOut.value = true
            }
            // BOTH branches still count toward the breaker
            // (RecoveryController.swift:272).
            recordAutoRecoveryOutcome(success = false, userInitiated = userInitiated)
            return
        }

        recordAutoRecoveryOutcome(success = true, userInitiated = userInitiated)
        if (generation != recoveryGeneration) return
        fetchSessions()
    }

    /**
     * Cancels any in-flight recovery and clears recovery UI state. Bumps the
     * generation so in-flight tasks bail at their next checkpoint, and suspends
     * the breaker so the next send failure doesn't immediately re-enter auto
     * recovery (RecoveryController.swift:284-300).
     */
    fun cancel() {
        recoveryGeneration += 1u
        recoveryJob?.cancel()
        recoveryJob = null
        _isRecovering.value = false
        // The cancelled pass's inner `finally` is generation-guarded (I2), so it
        // will NOT run its own `suppressSends(false)` (its captured generation no
        // longer matches the just-bumped value). Swift relies on the cancelled
        // Task's UNGATED defer to undo suppression; with the Android guard that
        // cleanup must happen here so a cancel can never strand
        // `suppressAllViewModelSends(true)`.
        suppressSends(false)
        isRecoveryDispatched = false
        recoveryFailed = true
        val now = nowMs()
        lastRecoveryEndedAt = now
        lastCancelledAt = now
        _autoRecoverySuspended.value = true
    }

    /**
     * Whether the recovery coroutine is still active (mirrors the Swift
     * `!Task.isCancelled` checkpoint at RecoveryController.swift:179).
     */
    private suspend fun currentlyActive(): Boolean = currentCoroutineContext().isActive

    // MARK: - Test hooks

    /** Read of the internal breaker failure counter (test parity assertions). */
    internal val consecutiveAutoFailuresForTest: Int get() = consecutiveAutoRecoveryFailures

    /** Seeds the breaker state for tests (Swift `_testOnly_setAutoRecoverySuspended`). */
    internal fun setBreakerForTest(suspended: Boolean, failures: Int) {
        _autoRecoverySuspended.value = suspended
        consecutiveAutoRecoveryFailures = failures
    }
}

File: Sources/ClaudeRelayClient/ViewModels/RecoveryController.swift
import Foundation
import os.log
import ClaudeRelayKit

private let recoveryLog = Logger(subsystem: "com.claude.relay.client", category: "Recovery")

/// Owns the recovery state machine: auto-suspend circuit breaker, generation
/// tokens, cooldown tracking, and the `handleForegroundTransition` /
/// `restoreSession` flow. Extracted from `SharedSessionCoordinator` in the
/// v0.3.2 hardening follow-up so the coordinator can stay under the SwiftLint
/// type-body ceiling without losing call-site ergonomics (all public entry
/// points on `SharedSessionCoordinator` forward to the controller unchanged).
///
/// The controller holds an `unowned` reference back to the coordinator —
/// ownership flows coordinator → controller, and the controller never
/// outlives its coordinator.
///
/// Call-site invariant: every method is `@MainActor`-isolated, matching the
/// coordinator. Recovery state (`isRecovering`, `recoveryFailed`,
/// `recoveryPhase`, etc.) is intentionally kept on the coordinator because
/// SwiftUI binds to those `@Published` properties — collapsing them into a
/// value type stays deferred (see plan Task 3.5).
@MainActor
final class RecoveryController {

    private unowned let coordinator: SharedSessionCoordinator
    private let connection: RelayConnection

    // MARK: - Breaker + generation state

    /// Monotonic token bumped at the start of every recovery pass. A scheduled
    /// `onSendFailed`-triggered recovery only runs if the token it captured
    /// still matches — prevents a failed recovery from immediately queueing
    /// another.
    private var recoveryGeneration: UInt64 = 0
    /// Timestamp of the last recovery completion (success or failure). Used
    /// to enforce a cooldown on auto-triggered recoveries via `onSendFailed`.
    private var lastRecoveryEndedAt: Date = .distantPast
    /// Consecutive auto-triggered recovery failures. Reset on success or on
    /// any user-initiated recovery (foreground, network restored, explicit).
    private var consecutiveAutoRecoveryFailures = 0
    /// A socket connected within this many seconds is trusted as alive without a
    /// ping round-trip, so the foreground/launch transition won't reconnect it
    /// out from under the launch `fetchSessions`.
    static let freshConnectionWindow: TimeInterval = 6
    /// Minimum delay between `onSendFailed`-triggered recoveries, in seconds.
    private let autoRecoveryCooldown: TimeInterval = 3
    /// After this many back-to-back auto-recovery failures we stop responding
    /// to `onSendFailed` until an explicit user/foreground/network signal
    /// arrives.
    private let maxAutoRecoveryFailures = 3
    /// True when auto-retry has been circuit-broken. Cleared on explicit
    /// recovery entry.
    private var autoRecoverySuspended = false
    /// True once the server rejected the token as invalid. Unlike the
    /// failure-count breaker, this trips on the FIRST auth rejection and is
    /// cleared ONLY by user-initiated recovery (e.g. after re-pairing). This
    /// is what stops the silent reconnect storm against a known-bad token.
    private var authRejected = false
    /// Synchronous entry lock, distinct from `isRecovering` (which is set
    /// after `await isAlive()`). Prevents concurrent recovery dispatches
    /// from racing across the suspension point at the top of
    /// `handleForegroundTransition`.
    private var isRecoveryDispatched = false
    /// Timestamp set by `cancel`. `triggerUserRecovery` ignores calls within
    /// 1 s of a cancel to avoid sheet-dismiss → scenePhase → re-trigger.
    private var lastCancelledAt: Date = .distantPast

    init(coordinator: SharedSessionCoordinator, connection: RelayConnection) {
        self.coordinator = coordinator
        self.connection = connection
    }

    // MARK: - Breaker

    /// Clears the auto-recovery circuit breaker. No-op when the breaker is
    /// already idle so healthy steady-state traffic doesn't thrash.
    func resetAutoRecoveryBreaker() {
        guard autoRecoverySuspended || consecutiveAutoRecoveryFailures > 0 else { return }
        consecutiveAutoRecoveryFailures = 0
        autoRecoverySuspended = false
        recoveryLog.info("Auto-recovery breaker reset after healthy ping")
    }

    // MARK: - Entry points

    /// Auto-triggered recovery entry point (called from `onSendFailed`).
    /// Gated by: torn-down state, already-recovering, auto-suspend (circuit
    /// broken), and a cooldown since the last recovery ended. User-initiated
    /// recovery (foreground/network/explicit) goes through
    /// `handleForegroundTransition` directly, bypassing these gates — see
    /// `triggerUserRecovery`.
    func scheduleAutoRecovery() {
        guard !coordinator.isTornDown else { return }
        guard !coordinator.isPerformingHandshake else {
            recoveryLog.debug("scheduleAutoRecovery: handshake in flight, ignoring")
            return
        }
        guard !isRecoveryDispatched else {
            recoveryLog.debug("scheduleAutoRecovery: already dispatched, ignoring")
            return
        }
        guard !autoRecoverySuspended else {
            recoveryLog.info("scheduleAutoRecovery: auto-suspend active — awaiting user signal")
            return
        }
        guard !authRejected else {
            recoveryLog.info("scheduleAutoRecovery: token rejected — awaiting user re-pair")
            return
        }
        let elapsed = Date().timeIntervalSince(lastRecoveryEndedAt)
        guard elapsed >= autoRecoveryCooldown else {
            recoveryLog.debug("scheduleAutoRecovery: cooldown (\(elapsed, format: .fixed(precision: 2))s < \(self.autoRecoveryCooldown)s)")
            return
        }
        recoveryLog.info("scheduleAutoRecovery: queuing recovery attempt")
        isRecoveryDispatched = true
        coordinator.recoveryTask = Task { [weak self] in
            await self?.handleForegroundTransition(userInitiated: false)
        }
    }

    /// Explicit user-initiated recovery: foreground, network restored, QR
    /// rescan, etc. Clears the auto-suspend circuit breaker so auto-retry
    /// resumes after this attempt.
    ///
    /// Always routes through `handleForegroundTransition`, which probes the
    /// socket with a real ping (`isAlive()`) before deciding whether to
    /// reconnect. A previous fast-path here trusted `connection.state ==
    /// .connected` and called `fetchSessions()` directly — but on macOS sleep
    /// the system suspends the WebSocket without flipping `state` to
    /// `.disconnected`, so the fast-path silently skipped reconnect on wake.
    /// The overhead of one ping per foreground/wake hop is negligible
    /// compared to the user-visible cost of a missed reconnect.
    func triggerUserRecovery() {
        guard !coordinator.isTornDown else { return }
        // A handshake already IS "connect, authenticate, get the session list" —
        // the entirety of what recovery would do. Letting recovery in here is
        // precisely the cold-launch bug: `scenePhase → .active` fires while the
        // launch handshake is mid-flight, `forceReconnect()` swaps the socket,
        // and the in-flight `session_list` dies with it.
        guard !coordinator.isPerformingHandshake else {
            recoveryLog.debug("triggerUserRecovery: handshake in flight, ignoring")
            return
        }
        guard !isRecoveryDispatched else {
            recoveryLog.debug("triggerUserRecovery: already dispatched, ignoring")
            return
        }
        if Date().timeIntervalSince(lastCancelledAt) < 1 {
            recoveryLog.debug("triggerUserRecovery: within cancel debounce, ignoring")
            return
        }
        autoRecoverySuspended = false
        consecutiveAutoRecoveryFailures = 0
        authRejected = false
        isRecoveryDispatched = true
        coordinator.recoveryTask = Task { [weak self] in
            await self?.handleForegroundTransition(userInitiated: true)
        }
    }

    /// Arm the auth-rejected gate from outside the recovery flow (e.g. when
    /// the initial attach path observes an invalid token). Idempotent.
    /// Cleared by `triggerUserRecovery` when the user re-pairs.
    func markAuthRejected() {
        authRejected = true
    }

    // MARK: - Recovery flow

    /// - Parameter userInitiated: true when triggered by an explicit
    ///   user-intent signal (scenePhase active, network restored, manual
    ///   retry). Such signals clear the auto-suspend circuit breaker.
    ///   Auto-triggered calls (send failed) increment the failure counter
    ///   on loss.
    func handleForegroundTransition(userInitiated: Bool) async {
        defer { isRecoveryDispatched = false }
        guard !coordinator.isTornDown else { return }
        guard !coordinator.isPerformingHandshake else {
            recoveryLog.debug("handleForegroundTransition: handshake in flight, skipping")
            return
        }
        guard !coordinator.isRecovering else {
            recoveryLog.debug("handleForegroundTransition: already recovering, skipping")
            return
        }

        // A just-established socket (e.g. on cold launch, `scenePhase` flips to
        // `.active` right after connect()) is alive by construction — its first
        // ping/pong hasn't completed yet, so `isAlive()` would return false and
        // trigger a needless `forceReconnect()` that TEARS DOWN the launch
        // handshake mid-flight, leaving the pane empty. Trust a recently
        // connected socket and hand straight to the handshake. (Proven from
        // server logs: two auths ~3.5s apart at launch with a disconnect
        // between = this reconnect.)
        if connection.state == .connected,
           let connectedAt = connection.lastConnectedAt,
           Date().timeIntervalSince(connectedAt) < Self.freshConnectionWindow {
            recoveryLog.info("handleForegroundTransition: connection just established, handshaking")
            await coordinator.performHandshake(reason: .wake)
            return
        }

        // Ping first, rather than letting the handshake discover the problem:
        // after macOS sleep the socket is dead while `state` still reads
        // `.connected`, and the ping detects that in 5 s where the handshake's
        // auth would sit for its full 10 s timeout before retrying.
        let alive = await connection.isAlive()
        if alive {
            recoveryLog.info("handleForegroundTransition: connection alive, handshaking")
            await coordinator.performHandshake(reason: .wake)
            return
        }

        recoveryGeneration &+= 1
        let myGeneration = recoveryGeneration
        recoveryLog.info("Recovery start gen=\(myGeneration) userInitiated=\(userInitiated)")

        coordinator.recoveryPhase = .reconnecting
        coordinator.recoveryFailed = false
        coordinator.isRecovering = true
        coordinator.suppressAllViewModelSends(true)
        defer {
            coordinator.isRecovering = false
            coordinator.suppressAllViewModelSends(false)
            lastRecoveryEndedAt = Date()
        }

        // Extended from [0, 1, 2, 4] to accommodate launchd's typical respawn
        // latency. Coupled with `onHealthyPing`'s breaker reset, an immediate
        // post-reconnect ping clears the auto-suspend state, so longer total
        // backoff doesn't trap legitimate retries in the breaker.
        let delays: [UInt64] = [0, 1, 2, 4, 8, 15]
        var reconnected = false
        for (attempt, delay) in delays.enumerated() {
            guard !coordinator.isTornDown, myGeneration == recoveryGeneration, !Task.isCancelled else {
                recoveryLog.info("Recovery aborted during reconnect (gen=\(myGeneration))")
                return
            }
            if delay > 0 {
                do {
                    try await Task.sleep(for: .seconds(delay))
                } catch {
                    recoveryLog.info("Recovery cancelled during backoff (gen=\(myGeneration))")
                    return
                }
                guard !coordinator.isTornDown, myGeneration == recoveryGeneration else { return }
            }

            do {
                try await connection.forceReconnect()
                reconnected = true
                break
            } catch is CancellationError {
                recoveryLog.info("Recovery cancelled during forceReconnect (gen=\(myGeneration))")
                return
            } catch {
                recoveryLog.error("forceReconnect attempt \(attempt + 1) failed: \(error.localizedDescription, privacy: .public)")
                if attempt == delays.count - 1 {
                    guard !coordinator.isTornDown, myGeneration == recoveryGeneration else { return }
                    coordinator.recoveryFailed = true
                    coordinator.connectionTimedOut = true
                    recordAutoRecoveryOutcome(success: false, userInitiated: userInitiated)
                    return
                }
            }
        }

        guard reconnected, !coordinator.isTornDown, myGeneration == recoveryGeneration else { return }
        await restoreSession(generation: myGeneration, userInitiated: userInitiated)
    }

    /// Records the outcome of a recovery pass and updates the circuit breaker.
    private func recordAutoRecoveryOutcome(success: Bool, userInitiated: Bool) {
        if success {
            consecutiveAutoRecoveryFailures = 0
            autoRecoverySuspended = false
            recoveryLog.info("Recovery succeeded — counters reset")
            return
        }
        if userInitiated {
            recoveryLog.info("User-initiated recovery failed — not counting toward auto-suspend")
            return
        }
        consecutiveAutoRecoveryFailures += 1
        recoveryLog.info("Auto-recovery failure \(self.consecutiveAutoRecoveryFailures)/\(self.maxAutoRecoveryFailures)")
        if consecutiveAutoRecoveryFailures >= maxAutoRecoveryFailures {
            autoRecoverySuspended = true
            recoveryLog.error("Auto-recovery suspended after \(self.maxAutoRecoveryFailures) consecutive failures")
        }
    }

    /// Restores auth + active session on the current connection.
    func restoreSession(generation: UInt64, userInitiated: Bool) async {
        coordinator.recoveryPhase = .authenticating
        coordinator.sessionController?.resetAuth()
        do {
            let controller = try await coordinator.ensureAuthenticated()
            guard !coordinator.isTornDown, generation == recoveryGeneration else { return }
            if let activeId = coordinator.activeSessionId {
                coordinator.recoveryPhase = .resuming
                coordinator.terminalViewModels[activeId]?.resetForReplay()
                try await controller.resumeSession(id: activeId)
                guard !coordinator.isTornDown, generation == recoveryGeneration else { return }
                coordinator.wireTerminalOutput(to: activeId)
            }
        } catch is CancellationError {
            recoveryLog.info("restoreSession cancelled (gen=\(generation))")
            return
        } catch {
            recoveryLog.error("restoreSession failed (gen=\(generation)): \(error.localizedDescription, privacy: .public)")
            guard !coordinator.isTornDown, generation == recoveryGeneration else { return }
            coordinator.recoveryFailed = true
            // App-level errors leave the connection intact — user can retry without reconnecting
            if SharedSessionCoordinator.isApplicationLevelError(error) {
                if case SessionController.SessionError.authenticationFailed = error {
                    // Token rejected: stop auto-recovery entirely until the
                    // user re-pairs. Prevents the reconnect storm that trips
                    // the server's rate limiter.
                    authRejected = true
                }
                // Session no longer exists / invalid transition / etc. The
                // socket itself is fine — clear the active session and surface
                // a recoverable error. Don't tear the workspace down via
                // connectionTimedOut.
                if let activeId = coordinator.activeSessionId {
                    coordinator.evictTerminal(for: activeId)
                    coordinator.activeSessionId = nil
                }
                coordinator.sessionAttachError = coordinator.friendlyAttachErrorMessage(error)
                coordinator.sessionAttachFailed = true
            } else {
                coordinator.connectionTimedOut = true
            }
            recordAutoRecoveryOutcome(success: false, userInitiated: userInitiated)
            return
        }

        recoveryLog.info("restoreSession success (gen=\(generation))")
        recordAutoRecoveryOutcome(success: true, userInitiated: userInitiated)
        guard !coordinator.isTornDown, generation == recoveryGeneration else { return }
        // Same contract as every other entry: the pane is repopulated from the
        // server's token-scoped list, and a session another device took while we
        // were disconnected is announced here (we missed its live push).
        await coordinator.performHandshake(reason: .wake)
    }

    /// Cancels any in-flight recovery and clears recovery UI state. Bumps
    /// the generation so in-flight tasks bail at their next checkpoint.
    func cancel() {
        recoveryLog.info("cancelRecovery requested")
        recoveryGeneration &+= 1
        coordinator.recoveryTask?.cancel()
        coordinator.recoveryTask = nil
        coordinator.authCoordinator.cancelInFlight()
        coordinator.isRecovering = false
        isRecoveryDispatched = false
        coordinator.recoveryFailed = true
        let now = Date()
        lastRecoveryEndedAt = now
        lastCancelledAt = now
        // Explicit cancel means the user doesn't want us re-entering
        // auto-recovery immediately on the next send failure — let them
        // re-enter via scenePhase or a fresh network restore.
        autoRecoverySuspended = true
    }

    /// Called by `SharedSessionCoordinator.tearDown()` so any in-flight
    /// dispatch flag is cleared before the coordinator drops the task.
    func invalidate() {
        isRecoveryDispatched = false
    }

    // MARK: - Test hooks

    var _testOnly_autoRecoverySuspended: Bool { autoRecoverySuspended }
    var _testOnly_consecutiveAutoRecoveryFailures: Int { consecutiveAutoRecoveryFailures }
    var _testOnly_authRejected: Bool { authRejected }
    func _testOnly_setAuthRejected(_ value: Bool) { authRejected = value }

    func _testOnly_setAutoRecoverySuspended(_ suspended: Bool, failures: Int) {
        autoRecoverySuspended = suspended
        consecutiveAutoRecoveryFailures = failures
    }
}
